### PR TITLE
[web] retrieve hostElement for an implicit view

### PR DESCRIPTION
**Pull request by p-mazhnik**
This PR updates `ui_web.views.getHostElement` API to also return `hostElement` for implicit view.

Resolves https://github.com/flutter/flutter/issues/150075

### CODE REVIEW

File: lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart [hunk 1]
```diff
@@ -99,6 +99,10 @@ class FlutterViewManager {
     return _jsViewOptions[viewId];
   }
 
+  DomElement? getHostElement(int viewId) {
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart:102]: This method could use a bit of `/// Documentation`.

**Review by ditman — COMMENTED** [review]
@p-mazhnik thanks for fixing this, I think this API makes sense (even though we'd like to get rid of the implicit view eventually).

Could you create an issue describing what this is solving, for the readers of the future, and link it with this PR?

**Comment by p-mazhnik** [comment]
@ditman could you please re-review and assign a second reviewer?

**Review by ditman — APPROVED** [review]
This LGTM! Thanks @p-mazhnik!

File: lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart [hunk 2]
```diff
@@ -99,6 +99,11 @@ class FlutterViewManager {
     return _jsViewOptions[viewId];
   }
 
+  /// Returns the host element in which the Flutter view associated to [viewId] is embedded (if any).
+  DomElement? getHostElement(int viewId) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart:103]: Do we have a clear definition of "host element"? If not, then the docs seem recursive. `getHostElement` already implied that it returns some "host element". There are two candidates for what people might think "host element" refers to:

- The `<flutter-view>` element.
- The direct parent of `<flutter-view>`.

The docs should explain which one this function returns.

Comment (ditman) [reply lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart:103]: I'd expect the `hostElement` to be the **parent element of the flutter view**.

Comment (p-mazhnik) [reply lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart:103]: That is what we have for `embeddingStrategy.hostElement`:
https://github.com/flutter/engine/blob/928d34f86a36df59985799e33c4f6dd2a6a5dfe6/lib/web_ui/lib/src/engine/view_embedder/embedding_strategy/embedding_strategy.dart#L29-L30

Also in the [Flutter docs](https://docs.flutter.dev/platform-integration/web/initialization):
```
HTML Element into which Flutter renders the app. 
```

I think `rootElement` is used for the `<flutter-view>`.

Comment (p-mazhnik) [reply lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart:103]: Doc in the code can be more specific though. How about updating doc for the `embeddingStrategy.hostElement` to be the following:

```dart
/// The dom element in which the Flutter view is embedded. 
/// This element is the direct parent element of the <flutter-view> element.
```

And then for the current function just leave it like this:
```dart
/// Returns the dom element in which the Flutter view associated to [viewId] is embedded (if any).
```

Comment (ditman) [reply lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart:103]: @p-mazhnik is the "(if any)" true? I think there's always a "parent element of the flutter view" (unless viewId is wrong)

Comment (p-mazhnik) [reply lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart:103]: Ah, I meant that there might not be a Flutter view associated with the provided `viewId`. I see now how this could be misinterpreted. Let's just omit it:
```
Returns the dom element in which the Flutter view associated to [viewId] is embedded
```

File: lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart [hunk 3]
```diff
@@ -99,6 +99,11 @@ class FlutterViewManager {
     return _jsViewOptions[viewId];
   }
 
+  /// Returns the dom element in which the Flutter view associated to [viewId] is embedded.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart:102]: nit: s/dom/DOM/

File: lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart (hunk 2 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart:103]: The revised doc comment is clearer. Thank you! However, it is more important to get the docs right in `dart:ui_web`, which is part of the public API. I have some suggestions there.

File: lib/web_ui/lib/ui_web/src/ui_web/flutter_views_proxy.dart [hunk 4]
```diff
@@ -17,13 +17,15 @@ class FlutterViewManagerProxy {
   // The proxied viewManager instance.
   final FlutterViewManager _viewManager;
 
-  /// Returns the `hostElement` configuration value passed from JS when `viewId` was added.
+  /// Returns the host element for [viewId].
```
Comment (yjbanov) [inline lib/web_ui/lib/ui_web/src/ui_web/flutter_views_proxy.dart:20]: Add one empty line between paragraphs.

File: lib/web_ui/lib/ui_web/src/ui_web/flutter_views_proxy.dart [hunk 5]
```diff
@@ -17,13 +17,15 @@ class FlutterViewManagerProxy {
...
+  /// In a multi-view app, it is the `hostElement` value passed from JS when [viewId] was added.
+  /// In an implicit view app, it is either the provided `hostElement` or the <body> element.
```
Comment (yjbanov) [inline lib/web_ui/lib/ui_web/src/ui_web/flutter_views_proxy.dart:22]: The `implicitView`-based element embedding is superseded by multi-view, so I wouldn't bother documenting it. Instead, we will deprecate it and remove it so we don't have multiple ways to do the same thing.

As for multi-view, we don't want to call it "multi-view app". Instead, we'll have full-page mode and add-to-app mode, and being able to add multiple view in the add-to-app mode is simply one of many capabilities of that mode. I'd reword this paragraph to reflect that, such as:

```
/// In the full-page mode, the host element is the `<body>` element of the page
/// and the view is the one and only [PlatformDispatcher.implicitView].
///
/// In the add-to-app mode, the host element is the value of `hostElement`
/// provided when creating the view.
```

Comment (p-mazhnik) [reply lib/web_ui/lib/ui_web/src/ui_web/flutter_views_proxy.dart]: This is much better, updated!

**Review by yjbanov — APPROVED** [review]
lgtm

**Comment by ditman** [comment]
Let's go!
